torch/distributed/_sharded_tensor/ops/linear.py

```diff
@@ -257,17 +262,93 @@ def _handle_row_wise_sharding_tensor(
 
     # Return the partial local result.
-    return results
+    return PartialTensor(results)
+
+
+def _init_sharded_tensor_from_local_result(
+    sharded_tensor,
+    local_result,
+    tensor_shard_dim,
+    result_shard_dim,
+    world_size,
+    pg,
+):
+    """
+    Given a sharded tensor and local_result from an op on top of it. We want
+    to create a new sharded tensor from the local_result so that the the next
+    op can be performed on the basis of the new sharded tensor. This can seen
+    as the last step of the first phase of the Megatron-LM style model(tensor)
+    parallelism.
+
+    Args:
+        sharded_tensor: Sharded tensor which the op was performed on.
+        local_result: A tensor which is from the op performed on the local_shard of
+            the sharded_tensor.
+        tensor_shard_dim: Dim which the tensor is sharded on.
+        result_shard_dim: Dim which the new sharded tensor will be sharded on.
+        world_size: number of ranks.
+        pg (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+
+    Return: new sharded tensor from the local_result.
+    """
+    sharded_weight_metadata = sharded_tensor.local_shards()[0].metadata
+    current_offsets = [0] * len(local_result.size())
+    current_offsets[result_shard_dim] = sharded_weight_metadata.shard_offsets[tensor_shard_dim]
+    local_shard_metadata = ShardMetadata(
+        shard_offsets=current_offsets,
+        shard_sizes=list(local_result.size()),
+        placement=sharded_weight_metadata.placement,
+    )
+    local_shards = [Shard(local_result, local_shard_metadata)]
+    global_size = list(local_result.size())
+    global_size[result_shard_dim] = sharded_tensor.size(tensor_shard_dim)
+
+    new_st = ShardedTensor._init_from_local_shards(local_shards, tuple(global_size), process_group=pg)
+
+    # Manually set sharding_spec
+    new_st._sharding_spec = copy.deepcopy(sharded_tensor._sharding_spec)
+    new_st._sharding_spec.dim = result_shard_dim
+    return new_st
+
+
+class _BiasTensorNarrow(Function):
+    @staticmethod
+    def forward(ctx, world_size, start_pos, chunk_size, weight, pg, bias):
+        ctx.weight = weight
+        ctx.pg = pg
+        ctx.world_size = world_size
+        return torch.narrow(bias, 0, start_pos, chunk_size)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        results = []
+        for idx in range(ctx.world_size):
+            results.append(grad_output.clone())
+        return (None, None, None, None, None) + (
+            _result_distribute_with_col_rearrange(results, grad_output, ctx.world_size, ctx.weight, ctx.pg),
+        )
+
+
+class _BiasTensorPartial(Function):
+    @staticmethod
+    def forward(ctx, world_size, bias):
+        ctx.world_size = world_size
+        return torch.div(bias, world_size)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return (None, grad_output)
```

Comment:
Do we really need this? The backward part of this function is not doing anything useful and just forwarding the grad?

Reply:
This is really tricky, I found that if I don't use this one, the grad will be incorrect. I will try to debug more here.

Reply:
I think I find why this is needed. Like `_BiasTensorNarrow`, in the forward, we multiply 1/4 (1/world_size) but during backward, we don't need that. That's why we need to manually skip the div during backward.